### PR TITLE
✨ add ability to configure DHCP overrides

### LINKED ISSUE #1582: Support configuration to disable DHCP DNS servers
/kind feature

**Describe the solution you'd like**
As a user, I would like to toggle a setting (`ignoreDHCPNameservers`) that will cause nodes to ignore the nameservers that DHCP assigns, so that I have full control over my DNS servers for my cluster.

This can be done with netplan already through the `dhcp{4,6}-overrides.use-dns` configuration option. However this currently only has an effect on the `networkd` backend. [netplan docs](https://netplan.io/reference#:~:text=The%20dhcp4%2Doverrides%20and%20dhcp6,override%20the%20default%20DHCP%20behavior.&text=Default%3A%20true%20.,effect%20on%20the%20networkd%20backend.)

We would like to add a field to the `NetworkDeviceSpec` on the `VsphereMachineTemplate`, something like:
```
// IgnoreDHCPNameservers is a flag that indicates whether or not to use the
// DNS servers received from the DHCP server.
// If true the DNS servers from the DHCP server will be ignored and only
// the statically configured nameservers will be added to the devices DNS
// servers.
// If false the DNS servers from DHCP and any statically configured
// nameservers will be added to the devices DNS servers.
// Please note that this only works with the `networkd` backend for netplan.
// +optional
IgnoreDHCPNameservers bool `json:"ignoreDHCPNameservers,omitempty"`
```

An example simplified config without all fields
```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: VSphereMachineTemplate
metadata:
  name: my-cluster
  namespace: my-namespace
spec:
  template:
    spec:
      network:
        devices:
          - networkName: "my-favorite-network"
            dhcp4: true
            dhcp6: true
            nameservers: [8.8.8.8, 8.8.4.4, 2001:4860:4860::8888]
            ignoreDHCPNameservers: true
```

An example metadata output
```
instance-id: "test-vm"
local-hostname: "test-vm"
wait-on-network:
  ipv4: true
  ipv6: true
network:
  version: 2
  ethernets:
    id0:
      match:
        macaddress: "00:00:00:00:00"
      set-name: "ens192"
      wakeonlan: true
      dhcp4: true
      dhcp6: true
      dhcp4-overrides:
        use-dns: false
      dhcp6-overrides:
        use-dns: false
      nameservers:
        addresses:
        - 8.8.8.8
        - 8.8.4.4
        - 2001:4860:4860::8888
```

The resulting netplan yaml
```
network:
  ethernets:
    id0:
      dhcp4: true
      dhcp4-overrides:
        use-dns: false
      dhcp6: true
      dhcp6-overrides:
        use-dns: false
      match:
        macaddress: xx:xx:xx:xx:xx:xx
      nameservers:
        addresses:
        - 8.8.8.8
        - 8.8.4.4
        - 2001:4860:4860::8888
      set-name: eth0
      wakeonlan: true
  version: 2
```

We'd love to get any input on the design and resolve any concerns there might be.

**Anything else you would like to add:**
This is related to https://github.com/vmware-tanzu/tanzu-framework/issues/1103

We are happy to PR this change and we have one started so we could explore our design. We are struggling a bit with running the e2e tests on our own and could use any help there as well but that might be something better to drop into k8s slack.

### PR DESCRIPTION
**Pull request by adobley**
**What this PR does / why we need it**:

This PR adds the ability to configure DHCP overrides for vSphere machines. This is important for users who want to tweak behaviors of their networks, such as using only the DNS servers that they configured.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1582

**Special notes for your reviewer**:

Our end-to-end test isn't concerned with the functional change these settings make, but with checking that capv writes the metadata correctly. Many of these settings are highly environment specific, knowing IPs or other configurations that could be difficult to account for in CI/dev environments. The responsibility of translating the metadata to network configuration falls on the distribution which we did not want to test here.

**Release note**:

```release-note
Add ability to configure DHCP overrides
```

### CODE REVIEW

**Comment by srm09** [comment]
@adobley This needs a rebase. 
/ok-to-test
@yastij @vrabbi this PR could benefit from a review.

**Comment by srm09** [comment]
@adobley could you rebase this PR, I can get a review on this one today

**Comment by srm09** [comment]
/test pull-cluster-api-provider-vsphere-full-e2e
To run the new e2e test introduced.

**Comment by srm09** [comment]
Considering the special notes that you have added, I wonder if the end to end test does provide something concrete? Could we not write unit and/or controller tests to mimic this behavior?

File: test/e2e/dhcp_overrides_test.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	})
+})
+
+func getVMMetadata(vm types.VSphereVM) []byte {
+	finder := find.NewFinder(vsphereClient.Client, false)
+	dc, err := finder.Datacenter(ctx, vm.Spec.Datacenter)
+	Expect(err).NotTo(HaveOccurred())
+	finder.SetDatacenter(dc)
+
+	vmObj, err := finder.VirtualMachine(ctx, fmt.Sprintf("/%s/vm/%s/%s", vm.Spec.Datacenter, vm.Spec.Folder, vm.Name))
```
Comment (srm09) [inline test/e2e/dhcp_overrides_test.go:100]:
```suggestion
	vmObj, err := finder.VirtualMachine(ctx, vm.Name)
```
should be enough to find the VM.

File: test/e2e/dhcp_overrides_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+type Ethernet struct {
+	DHCP4Overrides DHCPOverrides `yaml:"dhcp4-overrides"`
+}
+
+type DHCPOverrides struct {
+	SendHostname *bool `yaml:"send-hostname"`
+}
+
+var _ = Describe("DHCPOverrides configuration test", func() {
```
Comment (srm09) [inline test/e2e/dhcp_overrides_test.go:53]: This is missing an AfterEach block for cleanup of the cluster. Checkout https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/8cbf21df6b1632b48cf81cdc732bf1e44dca51b8/test/e2e/anti_affinity_test.go#L39-L59

File: test/e2e/dhcp_overrides_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+				err := yaml.Unmarshal(metadata, guestInfoMetadata)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).NotTo(BeNil())
+				Expect(*guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).To(BeFalse())
+			}
+		})
+	})
+})
+
+func getVMMetadata(vm types.VSphereVM) []byte {
```
Comment (srm09) [inline test/e2e/dhcp_overrides_test.go:94]: rewrite the signature to send an error if the metadata is not found and assert on the error on line 83, where the function is called.

File: test/e2e/dhcp_overrides_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	finder := find.NewFinder(vsphereClient.Client, false)
+	dc, err := finder.Datacenter(ctx, vm.Spec.Datacenter)
+	Expect(err).NotTo(HaveOccurred())
+	finder.SetDatacenter(dc)
```
Comment (srm09) [inline test/e2e/dhcp_overrides_test.go:98]: There is a global finder initialized for the e2e tests, consider using that.

**Review by srm09 — CHANGES_REQUESTED** [review]
Lotta changes in the e2e test, I have added some references to other tests you can use to model this e2e.

File: pkg/util/machines_test.go [hunk 5]
```diff
@@ -382,6 +382,84 @@ network:
       wakeonlan: true
       dhcp4: true
       dhcp6: true
+`,
+		},
+		{
+			name: "dhcp4+dhcp6+dhcpOverrides",
```
Comment (srm09) [inline pkg/util/machines_test.go:388]: add an explicit assertion for when the values are not set, just to verify the default behavior.

**Comment by srm09** [comment]
Also if it is not too much, could you try adding a controller test as well.

File: test/e2e/dhcp_overrides_test.go (hunk 1 above)
Comment (adobley) [reply test/e2e/dhcp_overrides_test.go]: Updated, thanks!

File: test/e2e/dhcp_overrides_test.go (hunk 3 above)
Comment (adobley) [reply test/e2e/dhcp_overrides_test.go]: Updated, thanks!

File: pkg/util/machines_test.go (hunk 5 above)
Comment (adobley) [reply pkg/util/machines_test.go:388]: Updated, thanks!

**Comment by yastij** [comment]
Mostly lgtm pending tests passing, same ask as @srm09 regarding controller tests. Thanks for working on this!

File: test/e2e/dhcp_overrides_test.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+			list := getVSphereVMsForCluster(clusterName, namespace.Name)
+			for _, vm := range list.Items {
+				metadata, err := getVMMetadata(vm)
+				Expect(err).NotTo(HaveOccurred())
+				guestInfoMetadata := &GuestInfoMetadata{}
+				err = yaml.Unmarshal(metadata, guestInfoMetadata)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).NotTo(BeNil())
+				Expect(*guestInfoMetadata.Network.Ethernets["id0"].DHCP4Overrides.SendHostname).To(BeFalse())
```
Comment (srm09) [inline test/e2e/dhcp_overrides_test.go:98]: From the `dhcp-overrides.yaml`, the override is being set only on the CP machine template. Update the kustomization to add it to both the templates, or a better test is to add different properties for the CP and worker node machine template.
When asserting them, the main difference would be that the CP VM would have the `cluster.x-k8s.io/control-plane: ""` label. So writing an assertion based off that would be the best thing to do.

Comment (adobley) [reply test/e2e/dhcp_overrides_test.go:98]: Thanks! Updated!

**Comment by srm09** [comment]
The [DHCP override e2e test](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-vsphere/1615/pull-cluster-api-provider-vsphere-full-e2e/1592973058149191680#1:build-log.txt%3A804-864) passed in the recent run.

**Pull request merged**